src/coreComponents/linearAlgebra/interfaces/hypre/HypreVector.cpp

```diff
@@ -481,169 +342,70 @@ void HypreVector::write( string const & filename,
   {
     case LAIOutputFormat::NATIVE_ASCII:
     {
-      GEOSX_LAI_CHECK_ERROR( HYPRE_IJVectorPrint( m_ij_vector, filename.c_str() ) );
+      GEOSX_LAI_CHECK_ERROR( hypre_ParVectorPrint( m_vec, filename.c_str() ) );
       break;
     }
     case LAIOutputFormat::MATRIX_MARKET:
     {
-      if( globalSize() == 0 )
+      MPI_Comm const comm = getComm();
+      int const rank = MpiWrapper::commRank( comm );
+
+      // Write MatrixMarket header
+      if( rank == 0 )
       {
-        if( MpiWrapper::commRank( getComm() ) == 0 )
-        {
-          FILE * fp = std::fopen( filename.c_str(), "w" );
-          hypre_fprintf( fp, "%s", "%%MatrixMarket matrix array real general\n" );
-          hypre_fprintf( fp, "%d %d\n", 0, 1 );
-          std::fclose( fp );
-        }
+        std::ofstream os( filename );
+        GEOSX_ERROR_IF( !os, GEOSX_FMT( "Unable to open file for writing: {}", filename ) );
+        os << "%%MatrixMarket matrix array real general\n";
+        os << GEOSX_FMT( "{} {}\n", globalSize(), 1 );
       }
-      else
+
+      if( globalSize() > 0 )
       {
-        // Copy distributed parVector in a local vector on every process
-        // with at least one component
+        // Copy distributed parVector in a local vector on every process with at least one component
         // Warning: works for a parVector that is smaller than 2^31-1
-        hypre_Vector *vector;
-        vector = hypre_ParVectorToVectorAll( m_par_vector );
+        hypre_Vector * const fullVector = hypre_ParVectorToVectorAll( m_vec );
```

Comment:
Don't we need here something as in `HypreExport`? 

https://github.com/GEOSX/GEOSX/blob/a924d56d2887c449ab40b82220dc531785fdd9c7/src/coreComponents/linearAlgebra/interfaces/hypre/HypreExport.cpp#L171-L186

Reply:
We do (and if only we had a unit test for `write()`, I would have found that out myself)